blueprints/user.permissions.md

```diff
@@ -0,0 +1,286 @@
+filling in a Role ID on user creation/update from a Role name if an ID was
+unset (this will now be handled by the API itself, which will save a full
+HTTP request on each user creation/modification).
+
+#### Database Impact
+The new model for a Role does not allow a `null` description; a simple
+migration that coalesces existing `NULL` values to an empty string and adds a
+check constraint should be all that's actually required. No other immediate
+changes should be made - including dropping the now-unused `api_capability`
+table -, since old API versions will still need access to the deprecated
+columns. However, the foreign key constraint on the `role_capability` table
+that links a "Capability" name to a row in the `capability` table should be
+dropped, as that is no longer the source of truth for valid Permissions.
+
+Optionally, a migration should really be added to make a user's username
+`NOT NULL`, since that field is actually required by the API and many things
+will break if it's `NULL`.
+
+## Documentation Impact
+The new configuration option will need to be documented, documentation for
+removed API endpoints will itself need to be removed, documentation for the
+`/roles` and `/users` endpoints will need to be updated to reflect the new
+request and response structures, and as Permissions are implemented on each
+endpoint the Permissions it requires for various actions will need to be
+defined.
+
+## Testing Impact
+The most significant testing changes will need to be made to the Go Traffic Ops
+client integration tests, which should verify each endpoint's proper
+Permissions-based authorization. Traffic Portal functionality will also require
+the appropriate end-to-end tests.
+
+## Performance Impact
+No significant performance impact is expected, since the Permissions of a user
+are already queried at the time of servicing a request by every authenticated
+endpoint. Some negligible, constant time will need to be spend determining how
+to authorize an authenticated user.
+
+## Security Impact
+Careful consideration must be given to the design and implementation of each
+Permission. For example, this author believes a Permission named
+`do-secrets-things` that allows a Role unrestricted read and write access to
+any and all DNSSEC, SSL, URL Signature, and URI signing keys would be a poor
+design from a security standpoint. Permissions should be broad enough to
+encompass a single, well-defined purpose, and no more. In many cases, though,
+the existing "Capabilities" concepts will be good enough to build from (e.g.
+`delivery-services-write`, `cdns-read`).
+
+## Upgrade Impact
+There will be a database migration to run, but since the default configuration
+will be to ignore Permissions and just use Privilege Level, there isn't much in
+the way of upgrade impact immediately. The bigger step is getting ready for the
+_next_ upgrade, when Permissions stop being optional.
+
+## Operations Impact
```

Comment:
granted i read thru this pretty quick but do you want to mention anything about how the current roles (admin, ops, r/o, etc) will be given perms to match their current behavior as dictated by priv level?

Reply:
Hopefully that's mostly taken care of by re-using the `role_capability` table, but yeah, we'll need to be more explicit in case people have created other Roles.  Seems like the safest thing to do is to "round down" the amount of privilege a Role grants. If a Role has a priv_level that isn't found exactly in routes.go, then we'll need to give it the same capabilities as the next-lowest that is, because we can't know the intention of such levels.